GitUI/CommandsDialogs/SettingsDialog/Pages/ColorsSettingsPage.cs

```diff
@@ -171,96 +189,98 @@ private void ChkHighlightAuthored_CheckedChanged(object sender, EventArgs e)
             _NO_TRANSLATE_ColorHighlightAuthoredLabel.Enabled = chkHighlightAuthored.Checked;
         }
 
-        private void BtnTheme_Click(object sender, EventArgs e)
+        private void ComboBoxTheme_SelectedIndexChanged(object sender, EventArgs e)
         {
-            ThemeModule.ShowEditor();
+            if (UpdatingThemeSettings())
+            {
+                return;
+            }
+
+            BeginUpdateThemeSettings();
+            UseSystemVisualStyle = SelectedThemeId == ThemeId.Default;
+            EndUpdateThemeSettings();
         }
 
-        private void BtnResetTheme_Click(object sender, EventArgs e)
+        private void ChkUseSystemVisualStyle_CheckedChanged(object sender, EventArgs e)
         {
-            ThemeModule.Controller.ResetTheme();
-            chkUseSystemVisualStyle.Checked = true;
+            BeginUpdateThemeSettings();
+            EndUpdateThemeSettings();
         }
 
-        private void ComboBoxTheme_SelectedIndexChanged(object sender, EventArgs e)
+        private void BeginUpdateThemeSettings()
         {
-            var menu = (ComboBox)sender;
-            if (menu.SelectedIndex < 0)
-            {
-                return;
-            }
+            _updateThemeSettingsCounter++;
+        }
 
-            if (_syncingTheme)
-            {
-                return;
-            }
+        private bool UpdatingThemeSettings() =>
+            _updateThemeSettingsCounter > 0;
 
-            _syncingTheme = true;
-            try
+        private void EndUpdateThemeSettings()
+        {
+            int counter = --_updateThemeSettingsCounter;
+            if (counter < 0)
             {
-                ThemeModule.Controller.SetTheme((string)menu.SelectedItem);
+                throw new InvalidOperationException($"{nameof(EndUpdateThemeSettings)} must be called after {nameof(BeginUpdateThemeSettings)}");
             }
-            finally
+
+            if (counter == 0)
             {
-                _syncingTheme = false;
+                bool settingsChanged =
+                    UseSystemVisualStyle != ThemeModule.Settings.UseSystemVisualStyle ||
+                    SelectedThemeId != ThemeModule.Settings.Theme.Id;
+
+                lblRestartNeeded.Visible = settingsChanged;
+                chkUseSystemVisualStyle.Enabled = SelectedThemeId != ThemeId.Default;
             }
         }
 
-        private void ChkUseSystemVisualStyle_CheckedChanged(object sender, EventArgs e)
+        private struct FormattedThemeId
```

Comment:
Why do we need this? Why can't we use `ThemeId`?

Reply:
`ToString()` is overriden to tell `ComboBox` menu item text. Overriding ToString() directly in `ThemeId` struct seems inappropriate since it's `ColorSettingsPage`'s UI concern, so should be handled there.